### ISSUE TITLE
Read Parameters from Spoolman into Printer

### ISSUE THREAD

**Issue by alexr10** [Donkie/Spoolman#415]
Right now the only communication between Klipper/Mainsail and Spoolman is for getting the spool usage data and checking for Filament compatibility.

Now I would appreciate it, if there would be an possibility to get more data from Spoolman for the printing process.
Examples would be:

- Extruder Temp
- Hotbed Temp
- Chamber Temp
- Extrusion Factor

This values above are filament specific and should be the same on each printer.
Furthermore, there are some values that are printer and filament specific which somebody might also want to store in Spoolman (especially in an single printer setup):

- Retraction
- Pressure Advance

I am now asking if there is a possibility to get the data from the Spoolman API and save it as variables in Klipper (that can be used with macros).
Ideally the variables should be updated on every filament change.

Thanks in advance.

**Comment by atl285** [comment]
Does it make sense to use it in Klipper? In which way? The print parameters will be defined in the Slicer not in Klipper / printer firmware. The only way I see is, to to validate the slicer gcode by matching the values against the filament values, for example if the extruder temperature with the filament range. 🤔

**Comment by alexr10** [comment]
I would like to have one generic Slicer Profile and get all the filament parameters from Spoolman.

I already did something similar a few months back. I had one generic slicer profile and set the filament parameters with Gcode macros as variables.
When a print got started the parameters have been read from the variables and set accordingly.
This solutions worked without issues, but it scaled really bad.

Furthermore there are some parameters, that are Klipper specific and are only set in firmware (Pressure Advance, ...)

**Comment by atl285** [comment]
But if you have multiple different printers, the Klipper specific values (Pressure Advance,...) are different. So you have to store these values printer specific. This is not the main focus of Spoolman, I think. 🤔

**Comment by alexr10** [comment]
I understand what you want to say.

But the functionality would also be useful for other Filament specific values:

- Extruder Temp
- Hotbed Temp
- Chamber Temp
- Retraction

In my opinion it would be nice to have the functionality to get more data (other than spool usage), when an filament change was done

**Comment by atl285** [comment]
Yes, I would like to have filament values from vendor in the filament settings, like Extruder and Hotbed temperature ranges, print speeds, density, vicat temperature, cooling settings, in the filament dataset. But how should they are usable? I see to functions:

* verifing sent gcode on printer before printing
* import filament (profile) into slicer

**Comment by alexr10** [comment]
You could put placeholders in the slicer software for certain values during slicing and get the real values during printing  from klipper.

**Comment by bofh69** [comment]
I've written a script that generates Orca Slicer (and SuperSlicer) config files from Spoolman data:
https://github.com/bofh69/spoolman2slicer

It uses jinja2 templates so it can be customized for whatever extra fields you use. The included Orca Slicer templates use an extra field for pressure advance for example.

**Comment by bofh69** [comment]
@alexr10 I've created a moonraker agent that listens for changes of the active spool. When it happens it will call gcode macros with the data about the new spool. You just create a gcode macro like `_SPOOLMAN_SET_FIELD_<fieldname>` for each field you are interested in and it will be called with `VALUE=<value>`. Its available here: [spool2klipper](https://github.com/bofh69/spool2klipper).
